rosbag2_compression/include/rosbag2_compression/sequential_compression_reader.hpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_COMPRESSION__SEQUENTIAL_COMPRESSION_READER_HPP_
+#define ROSBAG2_COMPRESSION__SEQUENTIAL_COMPRESSION_READER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2_compression/base_decompressor_interface.hpp"
+#include "rosbag2_compression/compression_options.hpp"
+
+#include "rosbag2_cpp/converter.hpp"
+#include "rosbag2_cpp/reader_interfaces/base_reader_interface.hpp"
+#include "rosbag2_cpp/serialization_format_converter_factory.hpp"
+#include "rosbag2_cpp/serialization_format_converter_factory_interface.hpp"
+
+#include "rosbag2_storage/metadata_io.hpp"
+#include "rosbag2_storage/storage_factory.hpp"
+#include "rosbag2_storage/storage_factory_interface.hpp"
+#include "rosbag2_storage/storage_interfaces/read_only_interface.hpp"
+
+#include "visibility_control.hpp"
+#include "zstd_decompressor.hpp"
+
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2_compression
+{
+
+class ROSBAG2_COMPRESSION_PUBLIC SequentialCompressionReader
+  : public ::rosbag2_cpp::reader_interfaces::BaseReaderInterface
+{
+public:
+  explicit SequentialCompressionReader(
+    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory =
+    std::make_unique<rosbag2_storage::StorageFactory>(),
+    std::shared_ptr<rosbag2_cpp::SerializationFormatConverterFactoryInterface> converter_factory =
+    std::make_shared<rosbag2_cpp::SerializationFormatConverterFactory>(),
+    std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io =
+    std::make_unique<rosbag2_storage::MetadataIo>());
+
+  virtual ~SequentialCompressionReader();
+
+  void open(
+    const rosbag2_cpp::StorageOptions & storage_options,
+    const rosbag2_cpp::ConverterOptions & converter_options) override;
+
+  void reset() override;
+
+  bool has_next() override;
+
+  std::shared_ptr<rosbag2_storage::SerializedBagMessage> read_next() override;
+
+  std::vector<rosbag2_storage::TopicMetadata> get_all_topics_and_types() override;
+
+  /**
+   * Ask whether there is another database file to read from the list of relative
+   * file paths.
+   *
+   * \return true if there are still files to read in the list
+   */
+  virtual bool has_next_file() const;
```

Comment:
I know this is only the second time we're writing this pattern, but the sequential file URI logic is a decent amount of copy-pasted code. It may be worth a refactor to delegate functionality to a common sequential file handler.

Reply:
I agree that it's a lot of similar code.
The problem with the sequential reader is that some of the functionality is private and not exposed in its public API. More over that's for another PR.

If we find ourselves doing this again though, then yes [a refactor is in order](https://refactoring.guru/refactoring/when).

Reply:
Splitting `SequentialReader` may be a step in resolving Issue https://github.com/ros2/rosbag2/issues/213 since `SequentialCompressionReader` requires a metadata file. 

I think there will be refactoring in the future where `SequentialReader` may split again into a specialized form for ROS1 bagfile support. If that happens, I think it would be more beneficial to refactor then.

Reply:
Why can't this class inherit from the `SequentialReader`?